CoreFeatures.md

```diff
@@ -0,0 +1,181 @@
+  - to you under the Apache License, Version 2.0 (the
+  - "License"); you may not use this file except in compliance
+  - with the License.  You may obtain a copy of the License at
+  -
+  -   http://www.apache.org/licenses/LICENSE-2.0
+  -
+  - Unless required by applicable law or agreed to in writing,
+  - software distributed under the License is distributed on an
+  - "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  - KIND, either express or implied.  See the License for the
+  - specific language governing permissions and limitations
+  - under the License.
+  -->
+
+# Parquet Core Features
+
+This document lists the core features for each parquet-format release. This
+list is a subset of the features which parquet-format makes available.
+
+## Purpose
+
+The list of core features for a certian release makes a compliance level that
+the different implementations can tied to. If an implementation claims that it
+provides the functionality of a parquet-format release core features it must
+implement all of the listed features according the specification (both read and
+write path). This way it is easier to ensure compatibility between the
+different parquet implementations.
+We cannot and don't want to stop our clients to use any features that are not
+on this list but it shall be highlighted that using these features might make
+the written parquet files unreadable by other implementations. We can say that
+the features available in a parquet-format release (and one of the
+implementations of it) and not on this list are experimental.
+
+## Versioning
+
+This document is versioned by the parquet-format releases which follows the
+scheme of semantic versioning. It means that no feature will be deleted from
+this document under the same major version. (We might deprecate some, though.)
+Because of the semantic versioning if one implementation supports the core
+features of the parquet-format release `a.b.x` it must be able to read any
+parquet files written by implementations supporting the release `a.d.y` where
+`b >= d`.
+
+If a parquet file is written according to a released version of this document
+it might be a good idea to write this version into the field `compliance_level`
+in the thrift object `FileMetaData`.
+
+## Adding new features
+
+The idea is to only include features which are specified correctly and proven
+to be useful for everyone. Because of that we require to have at least two
+different implementations that are released and widely tested.
+
+## The "list"
+
+This list is based on the [parquet thrift file](src/main/thrift/parquet.thrift)
+where all the data structures we might use in a parquet file are defined.
+
+### File structure
+
+All of the required fields in the structure (and sub-structures) of
+`FileMetaData` must be set according to the specification.
+The following page types are supported:
+* Data page V1 (see `DataPageHeader`)
+* Dictionary page (see `DictionaryPageHeader`)
```

Comment:
it might make sense to also call out here status of data page v2.

Reply:
You mean as a note? I am currently not sure if data page v2 worth to be supported at all. It was never widely used and the benefits of it are more about using some more advanced encodings and not the v2 page header itself.
So, I am not sure about the status of v2.

Reply:
Doesn't the data page v2 header allow selective compression of individual pages?

Reply:
Based on parquet.thrift V2 only allows to select if a page is compressed or not. The compression codec is specified in `ColumnMetaData` so it can be set by column (and not page) for both V1 and V2.

Reply:
Well, if some pages don't compress well you can skip the CPU cost of compression for them (I suppose that was the rationale for adding this feature?).

Reply:
I guess so. It sounds reasonably. I hope someone will join to the review who have deeper knowledge on these.

If we think V2 data page header worth it I am not against adding to the list for the first or maybe a later release. But first, we should list all the pros/cons.

Reply:
I like the selective compression feature personally, it's a smart idea. The other DataPageV2 layout changes wrt rep/def levels make sense.

We haven't implemented it in Impala, but it would be worthwhile if we determine that it's going to be universally supported (chicken/egg problem - no point implementing if we can't turn it on).

Reply:
It looks like parquet-cpp added support - https://issues.apache.org/jira/browse/PARQUET-458 so maybe we're the odd ones out now. https://issues.apache.org/jira/browse/IMPALA-6433 is the Impala JIRA

@wesm  said "BTW V2 files are not considered suitable for production by the Parquet community" - https://issues.apache.org/jira/browse/PARQUET-458?focusedCommentId=16931914&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16931914

This is probably relevant - not sure what issues he's referring to though.

Reply:
I meant that they aren't widely supported, so if you use it you will find implementations that can't read your files (per the chicken/egg problem above)

Reply:
So, what would be the general approach here? Do we want to add V2 to the first release of core features so we expect all the implementations to support it or we leave it for a next release?

Reply:
"Stable" is unfortunately underdefined here. It certainly seems stable as far as the _format_ is concerned. It's not necessarily widely implemented, though both parquet-mr and Parquet C++ support it.

Reply:
Trino also supports reading V2 data pages. However, when we attempted to use it in our parquet writer, we discovered other engines failed to read it https://github.com/trinodb/trino/issues/6377
There was confusion about whether V2 is officially finalised https://github.com/trinodb/trino/issues/7953#issuecomment-872544273
I believe support for V2 pages in other engines has also improved since then.
If it's simply a matter of adoption, then it would help to have some clarity about it in the spec (like for example acknowledging it as a core feature here) so that someone implementing a parquet writer can be assured that writing V2 pages is not some experimental feature that can be deprecated from the format.